pkg/chunk/cache/memcached.go

```diff
@@ -86,6 +157,53 @@ func (c *Memcached) Fetch(ctx context.Context, keys []string) (found []string, b
 	return
 }
 
+func (c *Memcached) fetchKeysBatched(ctx context.Context, keys []string) (found []string, bufs [][]byte, missed []string, err error) {
+	resultsCh := make(chan *result)
+	batchSize := c.cfg.BatchSize
+
+	go func() {
+		for i, j := 0, 0; i < len(keys); i += batchSize {
+			batchKeys := keys[i:util.Min(i+batchSize, len(keys))]
+			c.inputCh <- &work{
+				keys:     batchKeys,
+				ctx:      ctx,
+				resultCh: resultsCh,
+				batchID:  j,
+			}
+			j++
+		}
+	}()
+
+	// Read all values from this channel to avoid blocking upstream.
+	numResults := len(keys) / batchSize
+	if len(keys)%batchSize != 0 {
+		numResults++
+	}
+
+	// We need to order found by the input keys order.
+	results := make([]*result, numResults)
+	for i := 0; i < numResults; i++ {
+		result := <-resultsCh
+		results[result.batchID] = result
+	}
+	close(resultsCh)
+
+	for _, result := range results {
+		// TODO(gouthamve): One call may fail while everything else succeeds. Put the
+		// failed call keys in missed then.
+		if result.err != nil {
+			err = result.err
+			continue
```

Comment:
I'm a bit worries about partial failures as it stands.  The calling code will still use found/bufs/missed even if you return error here, but in that case we don't update missed.  So I think we need to do this TODO before we can merge.

Reply:
Also, can you add a test for this?